core-models/src/main/scala/com/mesosphere/usi/core/models/PodSpec.scala

```diff
@@ -12,4 +17,63 @@ package com.mesosphere.usi.core.models
   * @param goal target goal of this pod. See [[Goal]] for more details
   * @param runSpec WIP the thing to run, and resource requirements, etc.
   */
-case class PodSpec(id: PodId, goal: Goal, runSpec: RunSpec)
+trait PodSpec {
+  def id: PodId
+  def goal: Goal
+  def runSpec: RunSpec
+}
+
+object PodSpec {
+
+  /**
+    * Verifies that every value in range is requested only once. Requesting same value multiple times would yield pod that will never be scheduled.
+    * E.g. requesting two ports 80 on one machine is just not possible to satisfy
+    * @param resourceRequirements resources we are validating
+    * @return true if provided range requirements are valid
+    */
+  private def validateRangeRequirements(resourceRequirements: Seq[ResourceRequirement]): Unit Or One[ErrorMessage] = {
+    val staticPorts = resourceRequirements.collect {
+      case RangeRequirement(requestedValues, _, _) => requestedValues
+    }.flatten.collect { case ExactValue(value) => value }
+
+    if (staticPorts.distinct.length != staticPorts.length) {
+      Bad(
+        One(
+          ValidationError(
+            s"Every value inside RangeResource can be requested only once. Requirement: ${staticPorts.mkString(",")}")))
+    } else {
+      Good(())
+    }
+  }
+
+  private def validateScalarRequirements(resourceRequirements: Seq[ResourceRequirement]): Unit Or One[ErrorMessage] = {
+    val invalidScalar = resourceRequirements.collect {
+      case s: ScalarRequirement => s
+    }.filter(s => s.amount < 0)
+
+    invalidScalar match {
+      case Nil => Good(())
```

Comment:
I can see why you don't like it. I think this should actually return the resources. `Unit Or One[...]` does not seem right to me. `Or` and `Every` should not be part of a validation method but part of factory methods.

Reply:
but how do you accumulate then? All accumulating method expect to work with `Good Or Bad`

Reply:
If you return `Option[ErrorMessage]` here I think you can do that without Scalactic with the same result